docs/developers/completable-future.md

```diff
@@ -0,0 +1,78 @@
+
+Introduced in Java 8, *CompletableFuture* implements the older *Future* and adds support for callbacks, mapping operations, and cleaner awaiting of completion. If you are using Java 9 or later, you can also use delays and timeouts, subclass *CompletableFuture* to change the default executor, and take advantage of more utility methods.
+
+### What happens if a method returns CompletableFuture?
+
+In that case, you need to accept that you have a CompletableFuture and *not* the Object itself. You should continue your logic
+inside a callback, using the `thenAccept` method. Let's use the example of `CompletableFuture<Thing>`:
+
+```java
+CompletableFuture<Thing> futureThing = // ...
+futureThing.thenAccept((thing) -> {
+  // now you have the Thing
+  System.out.println("The thing is " + thing);
+  System.out.println("This message appears after futureThing is completed");
+});
+```
+
+When the future is completed, all callbacks are triggered. Usually callbacks use lambdas, but this is no requirement.
+
+If you don't care about the value of the CompletableFuture when it completes, you can use the `thenRun` method.
+
+### How do I return a value that relies on a CompletableFuture?
+
+This is a little bit more complicated. Sometimes you need to return a result which involves the `Thing`, suppose we call it `ThingWrapper`. You have written a partially complete method, but you don't now how to return `ThingWrapper`.
+
+```java
+public ThingWrapper getThingWrapper() {
+  CompletableFuture<Thing> futureThing = // ...
+
+  // How do I get "theThing"?
+  return new ThingWrapper(theThing);
+}
+```
+
+**The Lazy Way**
+
+If you absolutely must return a `ThingWrapper` immediately, there is nothing you can do except block the current thread and wait for the future to complete. You do this `CompletableFuture.join()`.
+
+```java
+public ThingWrapper getThingWrapper() {
+  CompletableFuture<Thing> futureThing = // ...
+
+  Thing theThing = futureThing.join();
+  return new ThingWrapper(theThing);
+}
+```
+
+However, this approach has its drawbacks. You do not know how long `futureThing` may take to complete. Maybe it's a database request which takes seconds to finish because some user's server has a MySQL database on the other side of the world.
+
+**The Comprehensive Way**
+
+The other way is to change the return type of your method to `CompletableFuture<ThingWrapper>`, and use one of the mapping operations:
+
+```java
+public CompletableFuture<ThingWrapper> getThingWrapper() {
+  CompletableFuture<Thing> futureThing = // ...
+
+  CompletableFuture<ThingWrapper> result = futureThing.thenApply((theThing) -> {
+    return new ThingWrapper(theThing);
+  });
+  return result;
+}
+```
+
+Now, you have transformed a `CompletableFuture<Thing>` into a `CompletableFuture<ThingWrapper>`. This approach may require some refactoring. You might need to change the return types of other methods which use `getThingWrapper`.
```

Comment:
Let's make this flow a bit better, perhaps after the "ThingWrapper" we add "however, this" instead of starting a new sentence.

Reply:
Good idea.

I also need to find a better example than "Thing" and "ThingWrapper", since they sound meaningless.

Reply:
Object could do, or even Foo/Bar